lib/tty/progressbar/meter.rb

```diff
@@ -14,32 +14,40 @@ class Meter
       #
       # @api private
       def initialize(interval)
-        @samples  = Hash.new { |h, el| h[el] = [] }
         @interval = interval || 1 # 1 sec
-        @marker   = 0
-
-        @start_time     = Time.now.to_f
-        @last_sample_at = @start_time
+        start
       end
 
       # Start sampling timer
       #
       # @api public
       def start
-        @start_time     = Time.now.to_f
-        @last_sample_at = @start_time
+        @start_time = Time.now
+        @current    = 0
+        @samples    = [[@start_time, 0]]
+        @rates      = Array.new
+        @start_time
       end
 
       # Update meter with value
       #
       # @api public
       def sample(at, value)
-        if @interval < (at.to_f - @last_sample_at)
-          @marker += 1
+        @current += value
+
+        # Remove samples that are obsolete and add the new one
+        cutoff = at - @interval
+        while @samples.size > 1 && (@samples.first.first < cutoff)
+          @samples.shift
         end
-        @samples[@marker] << value
+        @samples << [at, @current]
 
-        @last_sample_at = at.to_f
+        # If we crossed a period boundary since @start_time, save the rate for
+        # {#rates}
+        period_index = ((at - @start_time) / @interval).floor
+        while period_index > @rates.size
```

Comment:
@piotrmurach: I can fix that one if you really want to ... it would just look horrible to my eyes, so I prefer waiting on your opinion ....